File: Chapter 03/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter03/section_3_3_3_lsp/before/Goalie.java
package com.grokkingsoftwarearchitecture.chapter03.section_3_3_3_lsp.before;

/**
 * ANTI-PATTERN: Violation of the Liskov Substitution Principle (LSP).
 * * ARCHITECTURE PROBLEM: LSP states that objects of a superclass should be 
 * replaceable with objects of its subclasses without breaking the application.
 * * This Goalie class claims to be a "Player", but it refuses to fulfill the 
 * "playFieldPosition()" contract set by the base class. Often, developers will 
 * throw an UnsupportedOperationException here. This means the Goalie CANNOT be safely 
 * substituted anywhere a Player is expected.
 */
public class Goalie extends Player {
    
    /**
     * Breaks the base class contract by refusing to play the field.
     */
    @Override
    public void playFieldPosition() {
        // 🚨 ARCHITECTURE WARNING: A goalie doesn't play the field! If the Coach 
        // calls this blindly, they get unexpected behavior or an outright crash.
        System.out.println("  [Goalie] I can't do that! I stay near the net and use my hands.");
    }
}

File: Chapter 03/C#/Section_3_3_3_LSP/Before/Goalie.cs
using System;

namespace Chapter03.LSP.Before;

/// <summary>
/// ANTI-PATTERN: Violation of the Liskov Substitution Principle (LSP).
/// 
/// ARCHITECTURE PROBLEM: LSP states that objects of a superclass should be 
/// replaceable with objects of its subclasses without breaking the application.
/// 
/// This Goalie class claims to be a "Player", but it refuses to fulfill the 
/// "PlayFieldPosition()" contract set by the base class. Often, developers will 
/// throw a NotImplementedException here. This means the Goalie CANNOT be safely 
/// substituted anywhere a Player is expected.
/// </summary>
public class Goalie : Player
{
    /// <summary>
    /// Breaks the base class contract by refusing to play the field.
    /// </summary>
    public override void PlayFieldPosition()
    {
        // 🚨 ARCHITECTURE WARNING: A goalie doesn't play the field! If the Coach 
        // calls this blindly, they get unexpected behavior or an outright crash.
        Console.WriteLine("  [Goalie] I can't do that! I stay near the net and use my hands.");
    }
}